pkg/projectrequestlimit/validation.go

```diff
@@ -0,0 +1,256 @@
+	if err != nil {
+		return response.WithInternalServerError(request, err)
+	}
+	maxProjects, hasLimit, err := p.maxProjectsByRequester(userName, instance)
+	if err != nil {
+		return response.WithInternalServerError(request, err)
+	}
+	if hasLimit && projectCount >= maxProjects {
+		return response.WithForbidden(request, fmt.Errorf("user %s cannot create more than %d project(s)", userName, maxProjects))
+	}
+
+	return response.WithAllowed(request)
+}
+
+func (p *projectRequestLimitValidator) isApplicable(request *admissionv1.AdmissionRequest) bool {
+	if request == nil {
+		return false
+	}
+	if request.Operation != admissionv1.Create {
+		return false
+	}
+	desiredGR := projectv1.Resource("projectrequests")
+	if request.Resource.Group != desiredGR.Group {
+		return false
+	}
+	if request.Resource.Resource != desiredGR.Resource {
+		return false
+	}
+	if request.SubResource != "" {
+		return false
+	}
+	return true
+}
+
+func (p *projectRequestLimitValidator) isExempt(instance *projectv1.ProjectRequestLimit) bool {
+	if len(instance.Limits) == 0 &&
+		(instance.MaxProjectsForServiceAccounts == nil || *instance.MaxProjectsForServiceAccounts == 0) &&
+		(instance.MaxProjectsForSystemUsers == nil || *instance.MaxProjectsForSystemUsers == 0) {
+		return true
+	}
+	return false
+}
+
+func (p *projectRequestLimitValidator) getConfig() (*projectv1.ProjectRequestLimit, error) {
+	// list, err := p.prlLister.List(labels.Everything())
+	// if err != nil {
+	// 	return nil, err
+	// }
+	// if len(list) == 1 {
+	// 	return list[0], nil
+	// }
+	// if len(list) == 0 {
+	// 	return nil, apierrors.NewNotFound(
+	// 		projectv1.Resource(Resource),
+	// 		"cluster",
+	// 	)
+	// }
+	// if len(list) > 1 {
+	// 	return nil, fmt.Errorf("too many ProjectResourceLimit resources exist")
+	// }
+	return p.prlLister.Get("cluster")
+}
+
+// maxProjectsByRequester returns the maximum number of projects allowed for a given user, whether a limit exists, and an error
+// if an error occurred. If a limit doesn't exist, the maximum number should be ignored.
+func (p *projectRequestLimitValidator) maxProjectsByRequester(userName string, limits *projectv1.ProjectRequestLimit) (int64, bool, error) {
+	// service accounts have a different ruleset, check them
+	if _, _, err := serviceaccount.SplitUsername(userName); err == nil {
+		if limits.MaxProjectsForServiceAccounts == nil {
+			return 0, false, nil
+		}
+
+		return *limits.MaxProjectsForServiceAccounts, true, nil
+	}
+
+	// if we aren't a valid username, we came in as cert user for certain, use our cert user rules
+	if reasons := uservalidation.ValidateUserName(userName, false); len(reasons) != 0 {
+		if limits.MaxProjectsForSystemUsers == nil {
+			return 0, false, nil
+		}
+
+		return *limits.MaxProjectsForSystemUsers, true, nil
+	}
+
+	user, err := p.userLister.Get(userName)
+	if err != nil {
+		return 0, false, err
+	}
+	userLabels := labels.Set(user.Labels)
+
+	for _, limit := range limits.Limits {
```

Comment:
What's the precedence when there are multiple limits with different label selectors. Here the first match will be used. Is that documented or will it be documented?

Reply:
Exactly as you say: _Selectors are evaluated in order. The first one matching the current user will be used to determine the maximum number of projects._ 

Source: [3.11 documentation](https://docs.openshift.com/container-platform/3.11/admin_guide/managing_projects.html#limit-projects-per-user)